### PR TITLE
Add elevation to routing evaluations

### PR DESCRIPTION
**Pull request by KDKasonde**
**Objective**
This PR aims to tackle issue #18 with the omission of allowing end users to toggle edge colouring by steepness.

**Changes**
Config will now accept a `dem` attribute which is the location of the `tif` file users want to extract elevations from, but this is all optional.

The `Edge` struct now has the new attributes:
- `forward_cost` the cost of travelling on an edge from `node1` to `node2`
- `backward_cost` the cost of travelling on an edge from `node2` to `node1`
- `slope` the slope returned for a given edge as a result of the input `tif` file
- `slope_factor` an optional tuple that holds the `slope_factor` for the `forward_cost` and `backward_cost`

The steepness is used when doing batch calculations in the [calculate_batch](https://github.com/KDKasonde/od2net/blob/bfb74c1df6afddfdaec0c6526d1b76aa1c021e40/od2net/src/plugins/cost.rs#L12) function which returns an `option` tuple of costs now.  

For now the forward and backward costs are calculated by multiplying a base cost by the slope_factor. For more details see [here](https://github.com/U-Shift/Declives-RedeViaria/blob/main/SpeedSlopeFactor/SpeedSlopeFactor.md).

**Breaking changes**
Users with external commands need to edit them to make sure they output a `tuple` of costs not a `float` now as a result of the changes.

### CODE REVIEW

File: od2net/src/plugins/cost.rs [hunk 1]
```diff
@@ -29,21 +29,37 @@ pub fn calculate_batch(cost: &CostFunction, input_batch: Vec<&Edge>) -> Vec<Opti
             .into_iter()
             .map(|e| generalized(e, params))
             .collect(),
-        CostFunction::ExternalCommand(command) => external_command(command, input_batch).unwrap(),
+        // Was not sure how to add slope_factor to external commands, I also
```
Comment (dabreegster) [inline od2net/src/plugins/cost.rs:32]: What you have now looks fine to me, though I haven't tried it out in an example yet. Old comment? Also, the `into_iter().collect()` below can be removed; it doesn't do anything (turns a vec into an iterator, then back into a vec)

**Review by dabreegster — CHANGES_REQUESTED** [review]
Thanks so much for this! It's a very impressive first PR in Rust. How has learning the language been so far?

A bunch of comments, mostly Rust style stuff. Let me know if anything's unclear (or if you don't have time to address them). And could you please run `cargo fmt` to auto-format everything?

**Comment by KDKasonde** [comment]
> Thanks so much for this! It's a very impressive first PR in Rust. How has learning the language been so far?
> 
> A bunch of comments, mostly Rust style stuff. Let me know if anything's unclear (or if you don't have time to address them). And could you please run `cargo fmt` to auto-format everything?

Cheers man it is very fun to learn, definitely think reading the book first helped me out ALOT, but I still have a lot of trouble with the different pointers it uses and the traits system - I probably suck at design if it is in Rust. I am also slower when using it, but also much more thoughtful with what I am writing so I consider that a win. I will carry on using it, just because it is a very nice language.

And sweet I will look through the changes and makes some updates now!

File: od2net/src/network/create_from_osm.rs [hunk 2]
```diff
@@ -97,12 +109,52 @@ impl Network {
             let output_batch = plugins::cost::calculate_batch(cost, input_batch);
             for (key, cost) in key_batch.into_iter().zip(output_batch) {
                 progress.inc(1);
-                self.edges.get_mut(&key).unwrap().cost = cost;
+                
+                let output_cost = if let Some((forward_cost, backward_cost)) = cost {
```
[the remark this answers is not in the document]
Comment (KDKasonde) [reply od2net/src/network/create_from_osm.rs:113]: This is very nice!

File: od2net/src/network/mod.rs [hunk 3]
```diff
@@ -116,14 +118,114 @@ pub struct Edge {
...
+        };
+        
+        slope_factor
+    }  
+
+    fn get_slope<R: Read + Seek + Send>(&self, elevation_data: &mut GeoTiffElevation<R>) -> Option<f64> {
+        let first_node = self.geometry[0].to_degrees();
+        let second_node = self.geometry[1].to_degrees();
+
+        let first_node_height = match elevation_data.get_height_for_lon_lat(first_node.0 as f32, first_node.1 as f32) {
```
[the remark this answers is not in the document]
Comment (KDKasonde) [reply od2net/src/network/mod.rs:191]: Sweet that simplifies things a lot, didn't realise you could use it like that!

File: od2net/src/network/mod.rs [hunk 4]
```diff
@@ -116,14 +118,114 @@ pub struct Edge {
...
+        
+        let forward_slope_factor = Edge::calculate_slope_factor(slope, length);
+        let backward_slope_factor = Edge::calculate_slope_factor(-slope, length);
+
+        Some((slope, (forward_slope_factor, backward_slope_factor)))
+    }
+    
+    fn calculate_slope_factor(slope: f64, length: f64) -> f64 {
+    
+        let g = match (slope, length) {
```
[the remark this answers is not in the document]
Comment (temospena) [reply od2net/src/network/mod.rs:155]: I am not familiar with this language, but it looks alright!
As I wrote [here](https://github.com/U-Shift/Declives-RedeViaria/blob/main/SpeedSlopeFactor/SpeedSlopeFactor.Rmd#L114), "there is probably a way to write this code[equation] cleaner..." :slightly_smiling_face: 

Curious to see this in action!

File: viewer/package-lock.json
[the remark this answers is not in the document]
Comment (KDKasonde) [reply viewer/package-lock.json:1]: I will see if I can revert it, it seems to have updated some packages - I probably ran something else 🤔

**Review by dabreegster — APPROVED** [review]
This looks great, thank you so much! I'll pick things up tomorrow and update the examples to use this.

**Comment by KDKasonde** [comment]
Thanks for the comments btw they were really helpful! Also for the examples, how will you add the `geotiff` file?

**Comment by dabreegster** [comment]
I'll change the `setup.py` scripts to download the file, probably.

I'll open a PR and CC you on any changes I make tomorrow

**Comment by KDKasonde** [comment]
> I'll change the `setup.py` scripts to download the file, probably.
> 
> I'll open a PR and CC you on any changes I make tomorrow

Please do man!

**Pull request merged**